### ISSUE TITLE
Support for adding/removing workers to/from cluster

### ISSUE THREAD

**Issue by kvinwang** [Phala-Network/phala-blockchain#1190]
# The problem
Phala workers can become unavailable for various reasons, leading to a reduction in the number of workers in a cluster.
So we need a way to add new workers to a running cluster.

# The proposal
For removing workers, we can simply send an mq message to the worker and let the worker panic.

This proposal mainly focuses on how to add workers.

We can simply add an on-chain API that does the simple thing of `adding worker pubkey to the cluster`.
The pseudocode could be:
```
fn cluster_add_worker(origin, cluster_id, worker_pubkey) {
    ensure_cluster_owner(origin)?;
    ensure_worker_is_registered(worker_pubkey)?;
    ensure_worker_isn't_attached_to_any_cluster(worker_pubkey)?;
    cluster_workers(cluster_id).insert(worker_pubkey)
}
```

For off-chain workers, we add two pRPCs. One is to dump the cluster snapshot, another is to load the state.

The pseudocode to dump a snapshot:
```
fn cluster_prepare_snapshot_for(receiver_worker_pubkey) {
    ensure_worker_belongs_the_cluster(receiver_worker_pubkey)?;
    let filename = make_filename(block_number(), cluster_id(), receiver_worker_pubkey);
    encrypt_cluster_state_to(filename, receiver_worker_pubkey, tmp_key())?;
}
```
Here, we first check if the receiver_worker exists in the cluster then dump the state and encrypt it with the ECDH protocol. The state would be saved on the disk. Then we can mount an endpoint `/snapshots` on the public Rocket HTTP server to download the snapshots.

The pseudocode to load a snapshot:
```
fn cluster_load_snapshot(encrypted_state) {
    let state = decrypt_with_worker_privkey(encrypted_state);
    verify_hash(state)?;
    if state.block_number <= self.block_number {
        return Err("Staled snapshot");
    }
    self.cluster_state_to_be_loaded = Some(state);
}

fn on_block_end() {
    if self.cluster_state_to_be_loaded && self.cluster_state_to_be_loaded.block_numer == self.block_number {
        let state = self.cluster_state_to_be_loaded.take();
        ensure_myself_is_in_the_cluster(state.cluster_id)?;
        self.load_cluster_state(state);
    }
}
```
Here we load the state into the new worker. Then apply the state when the new worker synced the block that taking the snapshot.

Overall steps to add a new worker to a cluster:
1. Set up a new worker, sync it to the latest block, and register it on the chain.  Then pause syncing.
2. Invoke cluster_add_worker to add the new worker to the target cluster.
3. Invoke off-chain pRPC `cluster_prepare_snapshot_for(new_worke_pubkey)` on one of the workers in the cluster.
4. Download the snapshot via HTTP endpoint `/snapshots` to the new worker's disk.
5. Invoke off-chain pRPC `load_cluster_snapshot` on the new worker.
6. Continue to sync the new worker.

# Unresovled problems
- if all workers go offline, then there is no place the get the state.
- Should the chain automatically update the gas/storage price for the cluster while adding/removing workers?

**Comment by h4x3rotab** [comment]
> if all workers go offline, then there is no place the get the state.

That's reasonable. Could dumping the states periodically be a solution?

**Comment by kvinwang** [comment]
> That's reasonable. Could dumping the states periodically be a solution?
The dumping state needs the new worker's public key for the ECDH encryption in this proposal. So it can not be dumped periodically.

**Comment by kvinwang** [comment]
Resolved by #1219
